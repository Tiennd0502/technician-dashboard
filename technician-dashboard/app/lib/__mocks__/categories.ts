export const CATEGORIES = [
  {
    id: '01',
    label: 'Sofa Cleaning',
    image: 'https://i.ibb.co/QcSmTHW/category01.webp',
  },
  {
    id: '02',
    label: 'Bathroom',
    image: 'https://i.ibb.co/KDQzZ3s/category02.web',
  },
  {
    id: '03',
    label: 'Best Control',
    image: 'https://i.ibb.co/jwsPTTt/category03.webp',
  },
  {
    id: '04',
    label: 'View All',
    image: 'https://i.ibb.co/KxsM943/category04.webp',
  },
];
