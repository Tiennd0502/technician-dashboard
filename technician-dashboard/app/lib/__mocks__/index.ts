export * from './table';
export * from './categories';
export * from './services';
