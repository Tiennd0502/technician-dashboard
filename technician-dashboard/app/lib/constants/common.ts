export const LOGO_URL = '/images/technician.svg';

export const AVATAR_DEFAULT = '/images/empty-avatar.svg';

export const CURRENCY_CHARACTERS = {
  EURO: 'â‚¬',
};

export const TRUNCATE_STYLE = {
  textOverflow: 'ellipsis',
  overflow: 'hidden',
  whiteSpace: 'nowrap',
};

export const NO_DATA = 'No data found';
