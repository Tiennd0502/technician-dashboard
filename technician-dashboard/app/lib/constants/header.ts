import { AVATAR_DEFAULT } from '.';

export const HEADER_INFO = {
  user: {
    name: 'Belle',
    avatar: AVATAR_DEFAULT,
  },
  title: 'Product & Services',
};
