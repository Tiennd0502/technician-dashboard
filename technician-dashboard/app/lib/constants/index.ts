export * from './common';
export * from './filter';
export * from './sidebar';
export * from './header';
export * from './rules';
export * from './messages';
export * from './routes';
export * from './status';
export * from './table';
export * from './categories';
export * from './services';
