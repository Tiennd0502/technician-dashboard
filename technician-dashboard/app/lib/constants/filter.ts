import { SORT_TYPE } from '@/lib/interfaces';

export const PRODUCT_LIMIT = 5;

export const DEFAULT_PRODUCT_FILTER = {
  name: '',
  sortBy: 'name',
  order: SORT_TYPE.Asc,
};
