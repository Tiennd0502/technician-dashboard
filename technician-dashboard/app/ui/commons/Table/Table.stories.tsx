import type { Meta, StoryObj } from '@storybook/react';

import Table from '.';
import { DEFAULT_PRODUCT_FILTER, PRODUCTS } from '@/lib/constants';
import { generateProductHeaderColumn } from '@/lib/utils/table';

const meta: Meta<typeof Table> = {
  title: 'Components/Table',
  component: Table,
  tags: ['autodocs'],
};

export default meta;

type Story = StoryObj<typeof Table>;

export const Default: Story = {
  args: {
    filter: DEFAULT_PRODUCT_FILTER,
    columns: generateProductHeaderColumn(),
    data: PRODUCTS,
  },
};
