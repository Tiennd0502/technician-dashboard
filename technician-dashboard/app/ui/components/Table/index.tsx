'use client';

import { memo } from 'react';
import { Text, Flex, Box, Button, Checkbox, useMediaQuery } from '@chakra-ui/react';
import { TriangleDownIcon, TriangleUpIcon } from '@chakra-ui/icons';

// Types
import { TableData, TableHeader, SORT_TYPE, Filter } from '@/lib/interfaces';

// Components
import { TRUNCATE_STYLE } from '@/lib/constants';

interface TableProps {
  filter: Filter;
  columns: TableHeader[];
  data: TableData[];
  onEdit?: (id: string) => void;
  onDelete?: (id: string) => void;
}

const Table = ({ filter: { sortBy, order }, columns, data, onEdit, onDelete }: TableProps) => {
  const [isLargeScreen] = useMediaQuery('(min-width: 768px)');

  return (
    <Flex gap='1' flexDirection='column' w='full'>
      {isLargeScreen && (
        <Flex bg='background.section.primary' borderRadius='md' h='50px' w='100%' role='list'>
          {columns?.map(({ key, label, isCheckbox, width, onSort, isAction = false }) =>
            isCheckbox ? (
              <Box minW={width} key={key}></Box>
            ) : (
              <Flex key={key} alignItems='center' gap='5' w={width}>
                <Text
                  variant='textSm'
                  {...(isAction && {
                    w: 'full',
                    textAlign: 'right',
                    marginRight: '50px',
                  })}
                >
                  {label}
                </Text>
                {onSort ? (
                  <Flex flexDirection='column' cursor='pointer'>
                    <TriangleUpIcon
                      w='4'
                      h='2'
                      {...(sortBy === key &&
                        order === SORT_TYPE.Asc && {
                          color: 'primary.500',
                        })}
                      onClick={() => onSort({ sortBy: key, order: SORT_TYPE.Asc })}
                    />
                    <TriangleDownIcon
                      w='4'
                      h='2'
                      onClick={() => onSort({ sortBy: key, order: SORT_TYPE.Desc })}
                      {...(sortBy === key &&
                        order === SORT_TYPE.Desc && {
                          color: 'primary.500',
                        })}
                    />
                  </Flex>
                ) : null}
              </Flex>
            ),
          )}
        </Flex>
      )}

      {data.length ? (
        data.map((item) => (
          <Flex
            key={item.id}
            h={isLargeScreen ? '50px' : 'auto'}
            borderWidth='1px'
            borderRadius='md'
            borderColor='secondary'
            alignItems={isLargeScreen ? 'center' : 'flex-start'}
            flexDirection={isLargeScreen ? 'row' : 'column'}
            p={isLargeScreen ? 'auto' : '4'}
          >
            <Flex w='100%' flexDirection={isLargeScreen ? 'row' : 'column'}>
              {columns.map(({ key, width, label, isCheckbox, isAction, customView }) =>
                isCheckbox ? (
                  <Flex
                    key={key}
                    minW={width}
                    alignItems='center'
                    justifyContent={isLargeScreen ? 'center' : 'flex-start'}
                  >
                    <Checkbox size='lg' isDisabled />
                  </Flex>
                ) : (
                  <Flex
                    w={isLargeScreen ? width : '100%'}
                    key={key}
                    alignItems='center'
                    gap='4'
                    overflow='hidden'
                    pr='2'
                  >
                    {!isLargeScreen && (
                      <Text variant='textMd' mr='auto' w='max-content'>
                        {label}:
                      </Text>
                    )}
                    {customView ? (
                      customView(item[key])
                    ) : isAction ? (
                      <>
                        {onEdit && (
                          <Button px='2' ml='auto' mr='4' onClick={() => onEdit(item.id)}>
                            Edit{' '}
                          </Button>
                        )}
                        {onDelete && (
                          <Button variant='outline' mr='2' px='2' onClick={() => onDelete(item.id)}>
                            Delete
                          </Button>
                        )}
                      </>
                    ) : (
                      <Text variant='textSm' {...TRUNCATE_STYLE}>
                        {item[key]}
                      </Text>
                    )}
                  </Flex>
                ),
              )}
            </Flex>
          </Flex>
        ))
      ) : (
        <Flex
          h='50px'
          alignItems='center'
          justifyContent='center'
          borderWidth='1px'
          borderRadius='md'
          borderColor='secondary'
        >
          <Text variant='textMd'>No data found</Text>
        </Flex>
      )}
    </Flex>
  );
};

export default memo(Table);
