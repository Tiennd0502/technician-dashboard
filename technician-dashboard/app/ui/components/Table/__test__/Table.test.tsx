import { fireEvent, render, screen } from '@testing-library/react';
import * as mediaQueryHooks from '@chakra-ui/react';

import { DEFAULT_PRODUCT_FILTER, NO_DATA } from '@/lib/constants';
import { PRODUCTS, PRODUCT_HEADER_COLUMNS } from '@/lib/__mocks__';

import Table from '..';
import { SORT_TYPE } from '@/lib/interfaces';

jest.mock('@chakra-ui/react', () => {
  const originalModule = jest.requireActual('@chakra-ui/react');

  return {
    __esModule: true,
    ...originalModule,
    useMediaQuery: jest.fn().mockImplementation(() => [true]),
  };
});

describe('Table test cases', () => {
  jest.spyOn(mediaQueryHooks, 'useMediaQuery').mockImplementationOnce(() => [true]);

  const props = {
    filter: DEFAULT_PRODUCT_FILTER,
    columns: PRODUCT_HEADER_COLUMNS,
    data: PRODUCTS,
  };

  test('should render correctly', () => {
    const { container } = render(<Table {...props} />);
    expect(container).toMatchSnapshot();
  });

  test('Should render enough header elements', () => {
    const { getByRole } = render(<Table {...props} />);
    const header = getByRole('list');

    expect(header.childNodes).toHaveLength(PRODUCT_HEADER_COLUMNS.length);
  });

  test('Should render empty header elements', () => {
    const { getByRole } = render(<Table data={[]} columns={[]} filter={DEFAULT_PRODUCT_FILTER} />);
    const header = getByRole('list');

    expect(header.childNodes).toHaveLength(0);
  });

  test('Should render empty body elements', () => {
    render(<Table data={[]} columns={PRODUCT_HEADER_COLUMNS} filter={DEFAULT_PRODUCT_FILTER} />);

    const text = screen.getByTestId('no-data');
    expect(text.innerHTML).toEqual(NO_DATA);
  });

  test('Should render empty data', () => {
    render(<Table data={[]} columns={PRODUCT_HEADER_COLUMNS} filter={DEFAULT_PRODUCT_FILTER} />);

    const text = screen.getByTestId('no-data');
    expect(text.innerHTML).toEqual(NO_DATA);
  });

  test('Should render sort asc click', () => {
    render(<Table {...props} />);
    const button = screen.getByTestId('sort-asc');
    fireEvent.click(button);

    expect(jest.fn()).toHaveBeenCalled();
  });

  test('Should render sort desc click', () => {
    render(<Table {...props} filter={{ ...DEFAULT_PRODUCT_FILTER, order: SORT_TYPE.Desc }} />);
    const button = screen.getByTestId('sort-desc');
    fireEvent.click(button);

    expect(jest.fn()).toHaveBeenCalled();
  });

  test('should render correctly with mobile device', () => {
    jest.spyOn(mediaQueryHooks, 'useMediaQuery').mockImplementationOnce(() => [false]);

    const { container } = render(<Table {...props} />);
    expect(container).toMatchSnapshot();
  });

  test('Should render edit click', () => {
    render(<Table {...props} />);
    const button = screen.getByTestId('btn-edit');
    fireEvent.click(button);

    expect(jest.fn()).toHaveBeenCalled();
  });

  test('calls onEdit when being clicked', () => {
    const onEdit = jest.fn();
    render(<Table {...props} onEdit={onEdit} />);
    const button = screen.getByTestId('btn-edit');
    fireEvent.click(button);

    expect(onEdit).toHaveBeenCalled();
  });

  test('calls onDelete when being clicked', () => {
    const onDelete = jest.fn();

    render(<Table {...props} onDelete={onDelete} />);
    const button = screen.getByTestId('btn-delete');
    fireEvent.click(button);

    expect(onDelete).toHaveBeenCalled();
  });
});
