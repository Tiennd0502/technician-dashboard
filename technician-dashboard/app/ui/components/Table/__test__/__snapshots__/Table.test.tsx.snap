// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table test cases should render correctly 1`] = `
<div>
  <div
    class="css-1xauz2z"
  >
    <div
      class="css-17i2wn"
      role="list"
    >
      <div
        class="css-c8uo9c"
      />
      <div
        class="css-16f0jog"
      >
        <p
          class="chakra-text css-0"
        >
          Product name
        </p>
        <div
          class="css-1gb1rig"
        >
          <svg
            class="chakra-icon css-11mwvlo"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12.8,5.4c-0.377-0.504-1.223-0.504-1.6,0l-9,12c-0.228,0.303-0.264,0.708-0.095,1.047 C2.275,18.786,2.621,19,3,19h18c0.379,0,0.725-0.214,0.895-0.553c0.169-0.339,0.133-0.744-0.095-1.047L12.8,5.4z"
              fill="currentColor"
            />
          </svg>
          <svg
            class="chakra-icon css-weiglo"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M21,5H3C2.621,5,2.275,5.214,2.105,5.553C1.937,5.892,1.973,6.297,2.2,6.6l9,12 c0.188,0.252,0.485,0.4,0.8,0.4s0.611-0.148,0.8-0.4l9-12c0.228-0.303,0.264-0.708,0.095-1.047C21.725,5.214,21.379,5,21,5z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <div
        class="css-ty3ry7"
      >
        <p
          class="chakra-text css-0"
        >
          Brand name
        </p>
      </div>
      <div
        class="css-p3ab7x"
      >
        <p
          class="chakra-text css-0"
        >
          Service
        </p>
      </div>
      <div
        class="css-1kj85ga"
      >
        <p
          class="chakra-text css-0"
        >
          Status
        </p>
      </div>
      <div
        class="css-1n3sooa"
      >
        <p
          class="chakra-text css-1rrgtke"
        >
          Actions
        </p>
      </div>
    </div>
    <div
      class="css-zfiscw"
    >
      <div
        class="css-kzcie9"
      >
        <div
          class="css-13f3c3a"
        >
          <label
            class="chakra-checkbox css-1elxtiq"
            data-disabled=""
          >
            <input
              aria-disabled="true"
              class="chakra-checkbox__input"
              disabled=""
              style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
              type="checkbox"
              value=""
            />
            <span
              aria-hidden="true"
              class="chakra-checkbox__control css-dnty2r"
              data-disabled=""
            />
          </label>
        </div>
        <div
          class="css-sspw3w"
        >
          <p
            class="chakra-text css-6pogpz"
          >
            Washing machine
          </p>
        </div>
        <div
          class="css-17pip8y"
        >
          <p
            class="chakra-text css-6pogpz"
          >
            Hitachi
          </p>
        </div>
        <div
          class="css-ifuqlw"
        >
          <p
            class="chakra-text css-6pogpz"
          >
            Repair
          </p>
        </div>
        <div
          class="css-1490dfh"
        >
          <div
            class="css-70qvj9"
          >
            <div
              class="chakra-icon css-19b2uvg"
              focusable="false"
            />
            <p
              class="chakra-text css-0"
              data-testid="status-value"
            >
              Activated
            </p>
          </div>
        </div>
        <div
          class="css-4khqpw"
        >
          <div
            class="css-1gqs4bv"
          />
        </div>
      </div>
    </div>
    <div
      class="css-zfiscw"
    >
      <div
        class="css-kzcie9"
      >
        <div
          class="css-13f3c3a"
        >
          <label
            class="chakra-checkbox css-1elxtiq"
            data-disabled=""
          >
            <input
              aria-disabled="true"
              class="chakra-checkbox__input"
              disabled=""
              style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
              type="checkbox"
              value=""
            />
            <span
              aria-hidden="true"
              class="chakra-checkbox__control css-dnty2r"
              data-disabled=""
            />
          </label>
        </div>
        <div
          class="css-sspw3w"
        >
          <p
            class="chakra-text css-6pogpz"
          >
            Air conditioner
          </p>
        </div>
        <div
          class="css-17pip8y"
        >
          <p
            class="chakra-text css-6pogpz"
          >
            LG
          </p>
        </div>
        <div
          class="css-ifuqlw"
        >
          <p
            class="chakra-text css-6pogpz"
          >
            Installation
          </p>
        </div>
        <div
          class="css-1490dfh"
        >
          <div
            class="css-70qvj9"
          >
            <div
              class="chakra-icon css-crig9m"
              focusable="false"
            />
            <p
              class="chakra-text css-0"
              data-testid="status-value"
            >
              Deactivated
            </p>
          </div>
        </div>
        <div
          class="css-4khqpw"
        >
          <div
            class="css-1gqs4bv"
          />
        </div>
      </div>
    </div>
    <div
      class="css-zfiscw"
    >
      <div
        class="css-kzcie9"
      >
        <div
          class="css-13f3c3a"
        >
          <label
            class="chakra-checkbox css-1elxtiq"
            data-disabled=""
          >
            <input
              aria-disabled="true"
              class="chakra-checkbox__input"
              disabled=""
              style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
              type="checkbox"
              value=""
            />
            <span
              aria-hidden="true"
              class="chakra-checkbox__control css-dnty2r"
              data-disabled=""
            />
          </label>
        </div>
        <div
          class="css-sspw3w"
        >
          <p
            class="chakra-text css-6pogpz"
          >
            Smart TV
          </p>
        </div>
        <div
          class="css-17pip8y"
        >
          <p
            class="chakra-text css-6pogpz"
          >
            OnePlus
          </p>
        </div>
        <div
          class="css-ifuqlw"
        >
          <p
            class="chakra-text css-6pogpz"
          >
            Uninstallation
          </p>
        </div>
        <div
          class="css-1490dfh"
        >
          <div
            class="css-70qvj9"
          >
            <div
              class="chakra-icon css-19b2uvg"
              focusable="false"
            />
            <p
              class="chakra-text css-0"
              data-testid="status-value"
            >
              Activated
            </p>
          </div>
        </div>
        <div
          class="css-4khqpw"
        >
          <div
            class="css-1gqs4bv"
          />
        </div>
      </div>
    </div>
    <div
      class="css-zfiscw"
    >
      <div
        class="css-kzcie9"
      >
        <div
          class="css-13f3c3a"
        >
          <label
            class="chakra-checkbox css-1elxtiq"
            data-disabled=""
          >
            <input
              aria-disabled="true"
              class="chakra-checkbox__input"
              disabled=""
              style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
              type="checkbox"
              value=""
            />
            <span
              aria-hidden="true"
              class="chakra-checkbox__control css-dnty2r"
              data-disabled=""
            />
          </label>
        </div>
        <div
          class="css-sspw3w"
        >
          <p
            class="chakra-text css-6pogpz"
          >
            Water purifier
          </p>
        </div>
        <div
          class="css-17pip8y"
        >
          <p
            class="chakra-text css-6pogpz"
          >
            Kent
          </p>
        </div>
        <div
          class="css-ifuqlw"
        >
          <p
            class="chakra-text css-6pogpz"
          >
            Repair
          </p>
        </div>
        <div
          class="css-1490dfh"
        >
          <div
            class="css-70qvj9"
          >
            <div
              class="chakra-icon css-19b2uvg"
              focusable="false"
            />
            <p
              class="chakra-text css-0"
              data-testid="status-value"
            >
              Activated
            </p>
          </div>
        </div>
        <div
          class="css-4khqpw"
        >
          <div
            class="css-1gqs4bv"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
